python/ray/data/_internal/arrow_block.py

```diff
@@ -135,12 +138,29 @@ def __init__(self):
     @staticmethod
     def _table_from_pydict(columns: Dict[str, List[Any]]) -> Block:
         for col_name, col in columns.items():
-            if col_name == TENSOR_COLUMN_NAME or isinstance(
-                next(iter(col), None), np.ndarray
+            try:
+                if col_name == TENSOR_COLUMN_NAME or isinstance(
+                    next(iter(col), None), np.ndarray
+                ):
+                    from ray.data.extensions.tensor_extension import ArrowTensorArray
+
+                    columns[col_name] = ArrowTensorArray.from_numpy(col, col_name)
+                else:
+                    columns[col_name] = pyarrow.array(col)
+            except (
+                pyarrow.ArrowInvalid,
+                pyarrow.ArrowNotImplementedError,
+                pyarrow.ArrowTypeError,
```

Comment:
I believe this is the way that the fallback to Pandas is calculated; in any case, this code is outdated after the merge conflict because they have now all been coalesced into ArrowConversionError.

It is possible to check that all of the objects inside are of some Arrow type, but I believe this is quite cumbersome -- you have to manually check whether it is convertible to any PyArrow type, and you have to check all of the elements. It's easier to try conversion and see if it works.

In any case, the conversion into the ArrowPythonObjectArray should never fail if everything is pickleable, and I assume this is better than falling back to Pandas.